R/dependecies.R

```diff
@@ -0,0 +1,510 @@
+  names2 <- vapply(list2, function(obj) obj$name, character(1))
+
+  # Certain packages are no longer on cran but incorporated into R Core
+  core_pkgs <- as.character(installed.packages(priority = "base")[,1])
+  list1 <- list1[!names1 %in% core_pkgs]
+  names1 <- names1[!names1 %in% core_pkgs]
+  list2 <- list2[!names2 %in% core_pkgs]
+  names2 <- names2[!names2 %in% core_pkgs]
+
+  if (length(list1) == 0) return(list2)
+  if (length(list2) == 0) return(list1)
+
+  names(list1) <- names1
+  names(list2) <- names2
+  keep1 <- !names1 %in% names2
+
+  if (current_parent %in% names2 && any(names2 %in% names1)) {
+    init_parent_slot <- which(names2 == current_parent)
+    final_parent_slot <- max(which(names2 %in% names1))
+    if (final_parent_slot > init_parent_slot) {
+      sel1 <- seq_along(names2) != init_parent_slot & seq_along(names2) <= final_parent_slot
+      sel2 <- seq_along(names2) > final_parent_slot
+      list2 <- c(list2[sel1], list2[init_parent_slot], list2[sel2])
+      names2 <- c(names2[sel1], names2[init_parent_slot], names2[sel2])
+    }
+  }
+
+  list2 <- swap_versions(names1, names2, list1, list2)
+
+  c(list1[keep1], list2)
+}
+
+#' Swap all package information when side1 is greater than side2
+swap_versions <- function(names1, names2, list1, list2) {
+  swap_version2for1 <- vapply(
+    names1
+    , function(n) {
+      if (n %in% names2) {
+        obj1 <- list1[[n]]
+        obj2 <- list2[[n]]
+        if (obj1$is_dependency_package && obj2$is_dependency_package) {
+          if (is.na(obj1$version) && is.na(obj2$version)) {
+            if (obj1$remote != obj2$remote) {
+              package_version(obj1$latest_version) > package_version(obj2$latest_version)
+            } else {
+              FALSE
+            }
+          } else if(is.na(obj1$version)) {
+            FALSE
+          } else if(is.na(obj2$version)) {
+            TRUE
+          } else{
+            package_version(obj1$version) > package_version(obj2$version)
+          }
+        } else if (obj1$is_dependency_package){
+          FALSE
+        } else {
+          TRUE
+        }
+      } else{
+        FALSE
+      }}
+    , logical(1))
+  swap_versions1  <- names1 %in% names2 & swap_version2for1
+  swap_versions2 <- vapply(names1[swap_versions1]
+    , function(n) which(names2 == n)
+    , integer(1))
+  list2[swap_versions2] <- list1[swap_versions1]
+  list2
+}
+
+#' Either use the current lockbox library DESCRIPTION
+#' file or download the accurate remote DESCRIPTION file.
+get_dependencies <- function(package, lock) {
+  locked_package <- package
+  if (locked_package$is_dependency_package &&
+    !is.null(locked_package$latest_version_in_lockbox) &&
+    (is.na(locked_package$version) ||
+      package_version(as.character(locked_package$version)) <
+      package_version(as.character(locked_package$latest_version_in_lockbox)))) {
+      locked_package$version <- locked_package$latest_version_in_lockbox %||% NA
+  }
```

Comment:
Some indermediary assignment would actually be eludicating here, to explain what each conditional is testing.

Maybe they could even become functions.